packages/protocol/contracts/rrp/authorizers/Api3TokenLock.sol

```diff
@@ -0,0 +1,419 @@
+        );
+        require(
+            airnodeRequester.canUnlockAt[msg.sender] != 0,
+            ERROR_LOCK_PERIOD_ZERO
+        );
+        require(
+            airnodeRequester.canUnlockAt[msg.sender] <= block.timestamp,
+            ERROR_LOCK_PERIOD_NOT_EXPIRED
+        );
+        uint256 amount = airnodeRequester.lockAmountAt[msg.sender];
+
+        airnodeRequester.lockAmountAt[msg.sender] = 0;
+        airnodeRequester.canUnlockAt[msg.sender] = 0;
+        airnodeRequester.whitelistCount--;
+
+        if (airnodeRequester.whitelistCount == 0) {
+            IApi3RequesterRrpAuthorizer(api3RequesterRrpAuthorizer)
+                .setWhitelistExpiration(
+                    bytes32(abi.encode(_airnode)),
+                    _requesterAddress,
+                    0
+                );
+        }
+
+        assert(IApi3Token(api3Token).transfer(msg.sender, amount));
+
+        emit Unlock(_airnode, _requesterAddress, msg.sender, amount);
+    }
+
+    /// @notice User calls this when the lock amount has been decreased and wants
+    /// to withdraw the redundantly locked tokens
+    /// @param _airnode The airnode address
+    /// @param _requesterAddress The requester address
+    function withdraw(address _airnode, address _requesterAddress)
+        external
+        override
+        isNotBlacklisted(_airnode, _requesterAddress)
+    {
+        require(
+            airnodeToRequesterToTokenLocks[_airnode][_requesterAddress]
+                .lockAmountAt[msg.sender] > lockAmount,
+            ERROR_INSUFFICIENT_AMOUNT
+        );
+        uint256 withdrawAmount = airnodeToRequesterToTokenLocks[_airnode][
+            _requesterAddress
+        ].lockAmountAt[msg.sender] - lockAmount;
+        airnodeToRequesterToTokenLocks[_airnode][_requesterAddress]
+            .lockAmountAt[msg.sender] = lockAmount;
+        assert(IApi3Token(api3Token).transfer(msg.sender, withdrawAmount));
+
+        emit Withdraw(_airnode, _requesterAddress, msg.sender, withdrawAmount);
+    }
+
+    /// @notice Transfers the user's lock period and tokens to another requester
+    /// @param _airnode The airnode address
+    /// @param _fromRequesterAddress The requester from which the transfer will be done
+    /// @param _toRequesterAddress The target requester to which the transfer will be done
+    function transfer(
+        address _airnode,
+        address _fromRequesterAddress,
+        address _toRequesterAddress
+    )
+        external
+        override
+        isNotBlacklisted(_airnode, _fromRequesterAddress)
+        isNotBlacklisted(_airnode, _toRequesterAddress)
+    {
+        AirnodeRequester storage from = airnodeToRequesterToTokenLocks[
+            _airnode
+        ][_fromRequesterAddress];
+        AirnodeRequester storage to = airnodeToRequesterToTokenLocks[_airnode][
+            _toRequesterAddress
+        ];
+        require(
+            from.lockAmountAt[msg.sender] != 0,
+            "locked amount must be != 0"
```

Comment:
Should we also put this error msg in a constant along with the others?

Reply:
hmm yes, that would be better.